docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+
+We have ignored functions that use TF for now, but we will discuss those in the section devoted to point 3.
+
+### Single TF Session
+
+Given the single TF graph we have built for the current Swift for TF program, we can then go ahead and add the following in the beggining of the top-level Swift program code (on the Swift side):
+
+1. Create a new TF session (with any desirable configuration).
+2. Add a call to `session.run(sink)`.
+
+This last call will start execution of the TF program, so that it is running in parallel with the Swift program. It allows all parts of the graph that do not depend on communicating with Swift at any point in time, to be executed in parallel with the main Swift program. The fact that we end up with a single TF session means that it becomes easy to manage the session configuration and to profile the TF program performance. Furthermore, no two TF sessions have to fight for the same resources. It also makes this program slicing approach cleaner in that there is only two programs running at any point in time, a single Swift program and its corresponding TF program.
+
+## Tackling Point 2: Communication Between Swift and TF
+
+Communication between Swift and TF can be handled in a clean manner, by following the same principles.
+
+1. ___Swift -> TF:___ The TF program deals with `Tensor` objects. Therefore, whenever a `Tensor` object is constructed on the Swift side, and later used by TF ops that require tensors, a corresponding `SendToTF` op is added to the graph. This actually consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Dequeue` op will be used as the `Tensor` object representation for TF ops consuming it later on, and a `session.run(...)` call will be added to the Swift side, enqueueing the constructed tensor that will be sent to TF.
+2. ___TF -> Swift:___ `Tensor` objects allow for various ways to obtain their content, or a reference to them. One proposal could be to have methods like `.scalar`, `.toArray`, `.iterator`, and `.handle`, that return Swift-typed values. Whenever these are used, data needs to be transferred from TF to Swift and so we would add a `ReceiveFromTF` op to the graph. This actually again consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Enqueue` op will be added to the TF graph, enqueueing the corresponding TF tensor (or its handle, for example) to the queue. At the same time, a `session.run(...)` call will be added to the Swift side, dequeueing the TF content that is to be received from the graph. These method calls will be required to be made explicitly from the library user and will not be added automatically by the Swift compiler, in order to avoid confusion and potential inefficiencies.
```

Comment:
Hopefully you don't mean that the user will have to type "session.run" or something of the sort.

I also think this is missing discussion about how to just copy a pointer

Reply:
the design of Tensor pointer/handle can probably be decpupled from this doc. let's follow up on that separately.

Reply:
@alextp The compiler would add the call to `session.run`, and all other session management code. We should however provide a way for the user to pass their own `SessionConfig` if they want.

@mhong I agree. This document outlines very roughly some of the things we discuss about send/receive ops, but we should probably make these Moree concrete in a separate document. @alextp Regarding how to pass a pointer, there are a few ways we could do that, but the main thing we'd need to be careful about is reference counting when we pass tensor pointers around.

Reply:
Session config is very much like the existing target config. It can be passed either via a top-level method call, like `TensorFlow.enableGPU()`, or via a compiler flag, like `-tf-enable-tpu`.

Reply:
Oh I see. Session config has many more options though than currently supported in the existing target config though, right? I guess these could be supported by adding similar top-level method calls for them. We might also want to support directly loading a `SessionConfig` proto, as I can imagine people serializing configurations for each server, in a distributed setting.

Reply:
We need to be careful exposing proto-importing APIs for safety reasons, as loading a config proto at runtime won't provide early diagnostics.

A good first step is to add a compiler flag `-tf-session-config=<proto_file_path>` to compile the config proto along with the Swift program.